src/jltypes.c

```diff
@@ -70,93 +74,64 @@ jl_value_t *jl_nothing;
 
 // --- type properties and predicates ---
 
-STATIC_INLINE int is_unspec(jl_datatype_t *dt)
-{
-    return (jl_datatype_t*)dt->name->primary == dt;
-}
+typedef struct _typeenv {
+    jl_tvar_t *var;
+    jl_value_t *val;
+    struct _typeenv *prev;
+} jl_typeenv_t;
 
-static int jl_has_typevars__(jl_value_t *v, int incl_wildcard, jl_value_t **p, size_t np)
+static int typeenv_has(jl_typeenv_t *env, jl_tvar_t *v)
 {
-    size_t i;
-    if (jl_typeis(v, jl_tvar_type)) {
-        if (jl_has_typevars__(((jl_tvar_t*)v)->ub, incl_wildcard, p, np) ||
-            jl_has_typevars__(((jl_tvar_t*)v)->lb, incl_wildcard, p, np))
+    while (env != NULL) {
+        if (env->var == v)
             return 1;
-        if (p != NULL) {
-            for (i = 0; i < np; i++) {
-                if (v == p[i])
-                    return 1;
-            }
-            return 0;
-        }
-        if (!((jl_tvar_t*)v)->bound)
-            return incl_wildcard;
-        return 1;
-    }
-    if (jl_is_typector(v))
-        return incl_wildcard;
-    jl_svec_t *t;
-    int expect = -1;
-    if (jl_is_uniontype(v)) {
-        t = ((jl_uniontype_t*)v)->types;
-    }
-    else if (jl_is_datatype(v)) {
-        if (is_unspec((jl_datatype_t*)v))
-            return 0; // TODO: fix expect in this case
-        if (p == NULL) {
-            if (incl_wildcard)
-                expect = ((jl_datatype_t*)v)->haswildcard;
-            else
-                expect = ((jl_datatype_t*)v)->hastypevars;
-#ifdef NDEBUG
-            return expect;
-#endif
-        }
-        t = ((jl_datatype_t*)v)->parameters;
+        env = env->prev;
     }
-    else {
-        return 0;
+    return 0;
+}
+
+static int has_free_typevars(jl_value_t *v, jl_typeenv_t *env, int inv)
```

Comment:
comment re meaning of `inv` (invert?) and why it needs to be an argument (rather than applied by the caller)